File: 2023/go/day04.go
package main

import (
	"fmt"
	"strconv"
	"strings"
	"time"
)

type Scratchcard struct {
	id int
	reference map[int]bool
	obtained map[int]bool
}

func points(scratchcard Scratchcard) int {
	n_winning := len(intersection(scratchcard.reference, scratchcard.obtained))

	if n_winning == 0 {
		return 0
	}

	res := 1

	for i := 0; i < n_winning - 1; i++ {
		res *= 2
	}

	return res
}

func parse_scratchnumbers(numbers string) map[int]bool {
	res := make(map[int]bool)
	numbers = strings.TrimSpace(numbers)
	for _, str_number := range strings.Split(numbers, " ") {
		if len(str_number) == 0 {
			continue
		}
		number, err := strconv.Atoi(str_number)

		if err != nil {
			panic(err)
		}
		res[number] = true
	}

	return res
}

func parse_scratchcards(input string) []Scratchcard {
	res := make([]Scratchcard, 0)
	for _, line := range strings.Split(input, "\n") {
		parts := strings.Split(line, ":")

		card_id := strings.Split(parts[0], " ")
		id, err := strconv.Atoi(card_id[len(card_id) - 1])

		if err != nil {
			panic(err)
		}

		parts = strings.Split(parts[1], "|")

		reference := parse_scratchnumbers(parts[0])
		obtained := parse_scratchnumbers(parts[1])

		res = append(res, Scratchcard{
			id: id,
			reference: reference,
			obtained: obtained,
		})
	}

	return res
}

func day04_task01(scratchcards []Scratchcard) int {
	res := 0
	for _, scratchcard := range scratchcards {
		res += points(scratchcard)
	}

	return res
}

func day04_task02(scratchcards []Scratchcard) int {
	n_cards := make([]int, len(scratchcards))

	for _, scratchcard := range scratchcards {
		n_winning := len(intersection(scratchcard.reference, scratchcard.obtained))
		n_card := max(1, n_cards[scratchcard.id - 1])

		for i := 0; i < n_winning; i++ {
			if scratchcard.id + i >= len(n_cards) {
				break
			}
			n_cards[scratchcard.id + i] = max(1, n_cards[scratchcard.id + i]) + n_card
		}
	}

	res := 0
	for _, n_card := range n_cards {
		res += max(1, n_card)
	}

	return res
}

func day04_solution(input string) {
	input = strings.TrimSpace(input)

	start := time.Now()
	scratchcards := parse_scratchcards(input)
	elapsed := time.Since(start)
	fmt.Printf("Parse time: %s\n", elapsed)

	start = time.Now()
	fmt.Printf("Task 01: %d\n", day04_task01(scratchcards))
	elapsed = time.Since(start)
	fmt.Printf("Time: %s\n", elapsed)

	start = time.Now()
	fmt.Printf("Task 02: %d\n", day04_task02(scratchcards))
	elapsed = time.Since(start)
	fmt.Printf("Time: %s\n", elapsed)
}

File: 2023/python/day04.py
from dataclasses import dataclass, field
from pathlib import Path
from typing import Iterator

from loguru import logger
from typer import Typer
from utils import timer

main = Typer()


@dataclass
class Scratchcard:
    """Represents a scratchcard

    reference and obtained are sorted in ascending order.
    """

    id: int
    reference: list[int]
    obtained: list[int]

    _reference: list[int] = field(init=False, repr=False)
    _obtained: list[int] = field(init=False, repr=False)

    @classmethod
    def from_line(cls, line: str) -> "Scratchcard":
        """Parse input line into a Scratchcard"""
        line = line.strip()
        left, right = line.split(":")

        id = int(left.split(" ")[-1])

        ref, obt = right.split("|")
        return cls(
            id=id,
            reference=[int(x) for x in ref.split()],
            obtained=[int(x) for x in obt.split()],
        )

    @property
    def reference(self):
        return self._reference

    @reference.setter
    def reference(self, value: list[int]):
        self._reference = sorted(value)

    @property
    def obtained(self):
        return self._obtained

    @obtained.setter
    def obtained(self, value: list[int]):
        self._obtained = sorted(value)

    def winning_numbers(self) -> Iterator[int]:
        """Get the winning numbers for this scratchcard"""
        r_iter = iter(self.reference)
        o_iter = iter(self.obtained)

        try:
            ref = next(r_iter)
            obt = next(o_iter)
        except StopIteration:
            return

        while True:
            try:
                if ref == obt:
                    yield ref
                    ref = next(r_iter)
                    obt = next(o_iter)
                elif ref < obt:
                    ref = next(r_iter)
                else:
                    obt = next(o_iter)
            except StopIteration:
                break

    def points(self) -> int:
        """Compute the points for this scratchcard

        0 winning numbers: 0 points
        n>0 winning numbers: 2^(n-1) points
        """
        n_winnings = len(list(self.winning_numbers()))
        if n_winnings == 0:
            return 0

        return 2 ** (n_winnings - 1)


@timer
def task01(input: str) -> int:
    """Solution for task 01"""
    return sum(Scratchcard.from_line(line).points() for line in input.splitlines())


@timer
def task02(input: str) -> int:
    """Solution for task 02"""
    scratchcards = [Scratchcard.from_line(line) for line in input.splitlines()]
    # Contains the total number of scratchcards that can be obtained with
    total_cards: list[int] = [1] * len(scratchcards)

    for scratchcard in scratchcards:
        n_winnings = len(list(scratchcard.winning_numbers()))
        for idx in range(
            scratchcard.id, min(scratchcard.id + n_winnings, len(total_cards))
        ):
            total_cards[idx] += total_cards[
                scratchcard.id - 1  # Caution: card ids are 1-indexed
            ]

    return sum(total_cards)


@main.command()
def entrypoint(path: Path):
    """Entrypoint

    Args:
        path: Path to input file
    """
    with open(path, "r") as f:
        input = f.read().strip()

    logger.info(f"Task 01: {task01(input)}")
    logger.info(f"Task 02: {task02(input)}")


if __name__ == "__main__":
    main()
